### PR TITLE
ES2017 Rewrite

### LINKED ISSUE #12: Use atom-linter helpers for parsing and remove xregexp dependency
The `atom-linter` package now provides a `rangeFromLineNumber` function, which should remove the benefit of a custom parsing function.

Remove the custom parsing code and xregexp dependency in favor of shared code from `atom-linter`.

See also:   #11.

### LINKED ISSUE #13: Use atom-linter helpers for dealing with temp files
The `atom-linter` package intends to implement functionality for dealing with temp files.  When that happens, `linter-haml` should use that instead of custom code.

Depends on: AtomLinter/atom-linter#30.

See also: #11.

### LINKED ISSUE #16: Rubocop project config is not using by linter-haml
I have .rubocop.yml in root of my project, but it's not using in linter-haml.
I've defined line length, but it's not working.

### LINKED ISSUE #30: Plugin fails when .rubocop.yml inherits from another file
This happens if the option "Copy rubocop.yml to temporary directory" is active.

`linter-haml error Error: No such file or directory @ rb_sysopen - /var/folders/l1/c7gt0c8x1lsgn49kthq1fmyc0000gn/T/AtomLinter116021-27770-5yvtrn/.rubocop_todo.yml`

Here's how the .rubocop.yml file looks like:

```yaml
---
inherit_from: .rubocop_todo.yml

Style/DotPosition:
  EnforcedStyle: trailing

Style/MultilineOperationIndentation:
  EnforcedStyle: indented

Lint/EndAlignment:
  AlignWith: variable

Lint/HandleExceptions:
  Exclude:
    - 'bin/*'

```

### PR DESCRIPTION
**Pull request by Arcanemagus**
Major rewrite in ES2015, including:

* Clicking a rule name will now open the documentation
* Disable on the fly linting!
  * The file copying method used before wasn't complete enough to cover all edge cases and was just causing problems.
* Fix a race condition where file contents are changed after the lint is triggered
* Move configuration to `configSchema` in `package.json`

The new regex was validated here: https://regex101.com/r/EVwEiU/1

Fixes #30.
Fixes #16.
Closes #13.
Fixes #12.

### CODE REVIEW

**Comment by Arcanemagus** [comment]
@jschroeder9000 If you're still active here I'd love to have this looked over and your thoughts on the changes it is making.

File: spec/fixtures/.haml-lint.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+linters:
+  RuboCop:
+    enabled: true
+    ignored_cops:
+      - Style/FinalNewline
+      - Style/TrailingBlankLines
+      - Style/EndOfLine
```
Comment (Arcanemagus) [inline spec/fixtures/.haml-lint.yml]: For some reason this rule was triggering on all of the files for me, causing issues for the "valid" one. Unfortunately even though the other two are defined in the defaults and shouldn't be necessary here, it doesn't work without them.

File: package.json [hunk 2]
```diff
@@ -1,20 +1,30 @@
...
-  "main": "./lib/main",
+  "main": "./lib/main.js",
   "version": "1.2.1",
   "description": "Atom linter plugin for HAML, using haml-lint",
   "repository": "https://github.com/AtomLinter/linter-haml",
+  "private": true,
   "license": "MIT",
   "engines": {
-    "atom": ">=1.0.0 <2.0.0"
+    "atom": ">=1.4.0 <2.0.0"
```
Comment (Arcanemagus) [inline package.json]: Atom v1.4.0 was the first version to support the `configSchema`.

File: lib/main.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        if (hamlLintYmlPath) {
+          parameters.push('--config', hamlLintYmlPath);
+        } else if (globalHamlLintYmlFile !== '') {
+          parameters.push('--config', globalHamlLintYmlFile);
+        }
+
+        // Add the file to be linted
+        parameters.push(filePath);
+
+        return helpers.exec(executablePath, parameters, options).then((output) => {
```
Comment (steelbrain) [inline lib/main.js]: Can you make it an async function and await it instead? It'll save you from deeper nesting for so many lines

File: lib/main.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+          if (textEditor.getText() !== text) {
+            const warnMessage = 'File contents changed since the lint was triggered';
+            const warnOptions = {
+              description: 'The file was modified since the request was sent ' +
+                'to `linter-haml` to check it. Since any results would no longer ' +
+                'be valid, they are not being updated. Please save the file again ' +
+                'to update the results.',
+              dismissable: true,
+            };
+            atom.notifications.addWarning(warnMessage, warnOptions);
```
Comment (steelbrain) [inline lib/main.js]: This should be a console warning because this is too loud (as discussed privately)

File: lib/main.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            // The text has changed since the lint was triggered, tell Linter not to update
+            return null;
+          }
+
+          const messages = [];
+          let match = regex.exec(output);
```
Comment (steelbrain) [inline lib/main.js]: Code is duplicated here and at at line 94. You could make it like `while ((match = regex.exec(output)) !== null )`

File: package.json [hunk 6]
```diff
@@ -1,20 +1,30 @@
...
+  },
+  "configSchema": {
+    "hamlLintExecutablePath": {
```
Comment (steelbrain) [inline package.json]: Make it just `executablePath`

File: lib/main.js (hunk 5 above)
Comment (Arcanemagus) [inline lib/main.js]: That means you are assigning in the conditional of the `while`, which is against the code style guide.

File: package.json (hunk 6 above)
Comment (Arcanemagus) [inline package.json]: Since this will be a major release, that sounds like a good idea.

File: lib/main.js (hunk 5 above)
Comment (steelbrain) [inline lib/main.js]: I personally don't think it's wrong. Besides the code style guide was not given to us by God so we are free to object and change it to something that helps us keep the codebase clean and unduplicated

Comment (Arcanemagus) [inline lib/main.js]: Well, I personally agree with the code style guide so I'm keeping it the way it is :stuck_out_tongue:.

**Comment by Arcanemagus** [comment]
Giving this another day before merging.

**Comment by jschroeder9000** [comment]
Thanks for your work on this, @Arcanemagus.  I think this is a great start, though I find the decision to remove support for on the fly linting very disappointing.

I see that the linter is no longer using or even looking for a rubocop config at all.  That will severely cripple the usefulness of the linter by always linting with rubocop defaults instead of the what the user/project desires.

Haml-lint has added support for setting an environment variable to specify the path to the rubocop config file (`HAML_LINT_RUBOCOP_CONF`) and I believe that will solve our problems.  I think the right thing to do is to search for `.rubocop.yml` the same way the search for `.haml-lint.yml` is done and add the appropriate environment variable to the options passed to exec if it is found.  No file copying is necessary and I think this should also correctly solve #30.

With that implementation in place and since no file copying is being performed, I think on the fly linting should remain in place.

**Comment by jschroeder9000** [comment]
Open mouth, insert foot...

Haml-lint will find the `.rubocop.yml` and use it by itself since the source file is no longer being copied to a temporary location, so ignore my comments about needing to find it and pass an environment variable.  I guess the only reason we would need to pass an environment variable would be if we wanted to add an option for a global `.rubocop.yml` file like we do with `.haml-lint.yml`.  In fact, I don't think we need to be searching for a `.haml-lint.yml` file anymore either since files aren't being moved; haml-lint will find it by itself and use it.  We would only need to pass it in if the global config file is set.

Since there is no longer any file copying and that was the reasoning for disabling on the fly linting, I guess I'm confused as to why it is still being disabled.  I think the whole thing with hackily copying config files to a temp directory was a just a relic left over from the original linter plugin always copying files to a temp directory.

**Comment by Arcanemagus** [comment]
The searching for the file is left in to support the "global config file" option, otherwise we don't know if we need to specify our special one from the settings or not.

>Since there is no longer any file copying and that was the reasoning for disabling on the fly linting, I guess I'm confused as to why it is still being disabled.

I assumed from the fact that it was being copied in the first place that `haml-lint` doesn't support getting the contents from `stdin`, if we can send the current editor contents to it via `stdin` then of course we can just enable on the fly linting. I guess I never verified that it didn't 😛.

**Comment by jschroeder9000** [comment]
Ah, I had forgotten the issue with STDIN.  It's all coming back to me now.  Haml-lint does not support linting from STDIN and I agree simplifying things and disabling on the fly linting is the way to go.

I think this looks good and linking to the documentation is definitely a cool addition.

:+1: from me.

**Comment by Arcanemagus** [comment]
Published as v2.0.0! 🎉

**Pull request merged**